File: splash-screen/ios/Sources/SplashScreenPlugin/SplashScreenPlugin.swift
import Foundation
import UIKit
import Capacitor

@objc(SplashScreenPlugin)
public class SplashScreenPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "SplashScreenPlugin"
    public let jsName = "SplashScreen"
    public let pluginMethods: [CAPPluginMethod] = [
        .promise("show", SplashScreenPlugin.show),
        .promise("hide", SplashScreenPlugin.hide)
    ]
    private var splashScreen: SplashScreen?

    override public func load() {
        if let view = bridge?.viewController?.view {
            splashScreen = SplashScreen(parentView: view, config: splashScreenConfig())
            splashScreen?.showOnLaunch()
        }
    }

    // show and hide stay synchronous: the bridge queue runs them in the order of the calls and the splash screen hands
    // the UIKit work to the main queue in that order, so a hide after a show wins. Async methods would not keep that
    // order.

    // Show the splash screen
    public func show(_ call: CAPPluginCall) throws {
        guard let splash = splashScreen else {
            throw CAPPluginError("Unable to show Splash Screen")
        }
        let settings = splashScreenSettings(from: call)
        splash.show(settings: settings,
                    completion: {
                        call.resolve()
                    })
    }

    // Hide the splash screen
    public func hide(_ call: CAPPluginCall) throws {
        guard let splash = splashScreen else {
            throw CAPPluginError("Unable to hide Splash Screen")
        }
        let settings = splashScreenSettings(from: call)
        splash.hide(settings: settings)
        call.resolve()
    }

    private func splashScreenSettings(from call: CAPPluginCall) -> SplashScreenSettings {
        var settings = SplashScreenSettings()

        if let showDuration = call.getInt("showDuration") {
            settings.showDuration = showDuration
        }
        if let fadeInDuration = call.getInt("fadeInDuration") {
            settings.fadeInDuration = fadeInDuration
        }
        if let fadeOutDuration = call.getInt("fadeOutDuration") {
            settings.fadeOutDuration = fadeOutDuration
        }
        if let autoHide = call.getBool("autoHide") {
            settings.autoHide = autoHide
        }
        return settings
    }

    private func splashScreenConfig() -> SplashScreenConfig {
        var config = SplashScreenConfig()

        if let backgroundColor = getConfig().getString("backgroundColor") {
            config.backgroundColor = UIColor.capacitor.color(fromHex: backgroundColor)
        }
        if let spinnerStyle = getConfig().getString("iosSpinnerStyle") {
            switch spinnerStyle.lowercased() {
            case "small":
                config.spinnerStyle = .medium
            default:
                config.spinnerStyle = .large
            }
        }
        if let spinnerColor = getConfig().getString("spinnerColor") {
            config.spinnerColor = UIColor.capacitor.color(fromHex: spinnerColor)
        }
        config.showSpinner = getConfig().getBoolean("showSpinner", config.showSpinner)

        config.launchShowDuration = getConfig().getInt("launchShowDuration", config.launchShowDuration)
        config.launchAutoHide = getConfig().getBoolean("launchAutoHide", config.launchAutoHide)
        return config
    }

}

File: splash-screen/android/src/main/java/com/capacitorjs/plugins/splashscreen/SplashScreenPlugin.kt
package com.capacitorjs.plugins.splashscreen

import android.widget.ImageView
import com.getcapacitor.Logger
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.PluginThread
import com.getcapacitor.annotation.CapacitorPlugin
import com.getcapacitor.util.WebColor
import java.util.Locale

@CapacitorPlugin(name = "SplashScreen")
public class SplashScreenPlugin : Plugin() {
    private lateinit var splashScreen: SplashScreen
    private lateinit var splashScreenConfig: SplashScreenConfig

    override fun load() {
        splashScreenConfig = readSplashScreenConfig()
        splashScreen = SplashScreen(context, splashScreenConfig)
        if (!bridge.isMinimumWebViewInstalled() && bridge.config.errorPath != null && !splashScreenConfig.launchAutoHide) {
            return
        }
        splashScreen.showOnLaunch(activity)
    }

    // The splash screen's views and state belong to the main thread
    @PluginMethod(thread = PluginThread.MAIN)
    public fun show(call: PluginCall) {
        splashScreen.show(
            activity,
            getSettings(call),
            object : SplashListener {
                override fun completed() {
                    call.resolve()
                }

                override fun error() {
                    call.reject("An error occurred while showing splash")
                }
            }
        )
    }

    @PluginMethod(thread = PluginThread.MAIN)
    public fun hide(call: PluginCall) {
        if (splashScreenConfig.isUsingDialog) {
            splashScreen.hideDialog(activity)
        } else {
            splashScreen.hide(getSettings(call))
        }
        call.resolve()
    }

    override fun handleOnPause() {
        splashScreen.onPause()
    }

    override fun handleOnDestroy() {
        splashScreen.onDestroy()
    }

    private fun getSettings(call: PluginCall): SplashScreenSettings {
        val settings = SplashScreenSettings()
        call.getInt("showDuration")?.let { settings.showDuration = it }
        call.getInt("fadeInDuration")?.let { settings.fadeInDuration = it }
        call.getInt("fadeOutDuration")?.let { settings.fadeOutDuration = it }
        call.getBoolean("autoHide")?.let { settings.autoHide = it }
        return settings
    }

    private fun readSplashScreenConfig(): SplashScreenConfig {
        val splashConfig = SplashScreenConfig()
        val backgroundColor = config.getString("backgroundColor")
        if (backgroundColor != null) {
            try {
                splashConfig.backgroundColor = WebColor.parseColor(backgroundColor)
            } catch (ex: IllegalArgumentException) {
                Logger.debug("Background color not applied")
            }
        }
        splashConfig.launchShowDuration = config.getInt("launchShowDuration", splashConfig.launchShowDuration)
        splashConfig.launchFadeOutDuration = config.getInt("launchFadeOutDuration", splashConfig.launchFadeOutDuration)
        splashConfig.launchAutoHide = config.getBoolean("launchAutoHide", splashConfig.launchAutoHide)
        config.getString("androidSplashResourceName")?.let { splashConfig.resourceName = it }
        splashConfig.isImmersive = config.getBoolean("splashImmersive", splashConfig.isImmersive)
        splashConfig.isFullScreen = config.getBoolean("splashFullScreen", splashConfig.isFullScreen)

        val spinnerStyle = config.getString("androidSpinnerStyle")
        if (spinnerStyle != null) {
            splashConfig.spinnerStyle =
                when (spinnerStyle.lowercase(Locale.ROOT)) {
                    "horizontal" -> android.R.attr.progressBarStyleHorizontal
                    "small" -> android.R.attr.progressBarStyleSmall
                    "large" -> android.R.attr.progressBarStyleLarge
                    "inverse" -> android.R.attr.progressBarStyleInverse
                    "smallinverse" -> android.R.attr.progressBarStyleSmallInverse
                    "largeinverse" -> android.R.attr.progressBarStyleLargeInverse
                    else -> android.R.attr.progressBarStyleLarge
                }
        }
        val spinnerColor = config.getString("spinnerColor")
        if (spinnerColor != null) {
            try {
                splashConfig.spinnerColor = WebColor.parseColor(spinnerColor)
            } catch (ex: IllegalArgumentException) {
                Logger.debug("Spinner color not applied")
            }
        }
        val scaleTypeName = config.getString("androidScaleType")
        if (scaleTypeName != null) {
            splashConfig.scaleType =
                try {
                    ImageView.ScaleType.valueOf(scaleTypeName)
                } catch (ex: IllegalArgumentException) {
                    ImageView.ScaleType.FIT_XY
                }
        }

        splashConfig.showSpinner = config.getBoolean("showSpinner", splashConfig.showSpinner)
        splashConfig.isUsingDialog = config.getBoolean("useDialog", splashConfig.isUsingDialog)
        config.getString("layoutName")?.let { splashConfig.layoutName = it }

        return splashConfig
    }
}
